websocket/src/emscripten_callbacks.cpp

```diff
@@ -0,0 +1,57 @@
+#include "websocket.h"
+
+#if defined(__EMSCRIPTEN__)
+
+namespace dmWebsocket
+{
+EM_BOOL Emscripten_WebSocketOnOpen(int eventType, const EmscriptenWebSocketOpenEvent *websocketEvent, void *userData) {
+    DebugLog(1, "WebSocket OnOpen");
+    // WebsocketConnection* conn = FindConnectionForEmscriptenWebSocket(websocketEvent->socket);
+    WebsocketConnection* conn = (WebsocketConnection*)userData;
+    if (IsConnectionValid(conn))
```

Comment:
I don't really see why we need this extra function?
Don't we know what states the connection passes through? Can it be changed/deleted at any time?

Reply:
Well the problem is that we can call websocket.disconnect() while the JavaScript WebSocket connection initiated through Emscripten is in progress. We will in that case get the callback to Emscripten_WebSocketOnOpen() AFTER we have destroyed the connection in websocket.cpp and free'd the WebsocketConnection struct.

Reply:
Do they call the callback _after_ you've called "emscripten_websocket_close(conn->m_WS, 1000, "DestroyConnection");" ?

Reply:
No I wouldn't expect that no.

Reply:
And I see what you are getting at :-) Let me test again.

Reply:
Also, I think that if this _is_ needed, you need to wait for the callbacks to finish.
E.g. destroy the connection once everything is finished.
That way you could instead check the state of the connection itself, not the global state array of all connections.

Reply:
Sending a messages to an echo server (which will return the same message) and immediately closing the connection can result in:

    exception thrown: RuntimeError: memory access out of bounds,RuntimeError: memory access out of bounds

I recall reading about how send and close are handled when there are in and outbound messages. Need to find it in the spec.

Reply:
Log with debug = 2

```
DEBUG:SCRIPT: Connecting to wss://echo.websocket.org
WARNING:WEBSOCKET: STATE_CREATE -> STATE_CONNECTING
WARNING:WEBSOCKET: WebSocket OnOpen
WARNING:WEBSOCKET: STATE_CONNECTING -> STATE_CONNECTED
DEBUG:SCRIPT: Connected: userdata: 0x16361e8
DEBUG:SCRIPT: Sending 'sending to server', nil, nil, nil, nil
WARNING:WEBSOCKET: WebSocket OnMessage
WARNING:WEBSOCKET: PushMessage 'sending to server' 17 bytes
DEBUG:SCRIPT: Receiving: 'sending to server'
WARNING:WEBSOCKET: STATE_CONNECTED -> STATE_DISCONNECTED
DEBUG:SCRIPT: Disconnected: userdata: 0x16361e8
WARNING:WEBSOCKET: DestroyConnection: 0x16361e8
WARNING:WEBSOCKET: WebSocket OnClose
WARNING:WEBSOCKET: PushMessage '' 0 bytes
exception thrown: RuntimeError: memory access out of bounds,RuntimeError: memory access out of bounds
```

Reply:
That error doesn't really invalidate my concerns of using a global state to determine if this particular connection is being destroyed or not.

What callback is it calling there?

I think my point still stands: Don't delete the connection until you've shut it down (i.e. waited for emscripten to call it's callbacks (i.e. close).

Reply:
And I see now that the problem is with the OnClose callback and not sending and closing.

```
websocket.disconnect()
LuaDisconnect()
   CloseConnection()
       STATE_DISCONNECTED
   emscripten_websocket_close()

OnUpdate() STATE_DISCONNECTED
  DestroyConnection()

Emscripten_WebSocketOnClose()
   PushMessage()
      CRASH! (connection already destroyed)
```

Reply:
You can either wait for it. Or let that function actually delete the connection.

Reply:
I don't feel comfortable with that approach. It would mean that freeing memory in the engine is dependent upon JavaScript and the WebSocket API always generating a "close" event. But perhaps I'm overly cautious?

Reply:
Yeah, deleting in the callback is not the way I'd usually do it. And it's also dependent on how much you trust that api.

The alternative is as I mentioned to wait for it to disconnect. Which fits nicely with our state machine that handles cases like this. Perhaps a new Disconnecting state is in order?

Reply:
I'll rewrite the state handling a bit when disconnecting. Introducing a DISCONNECTING state will help.